packages/protocol/test/governance/epochrewards.ts

```diff
@@ -0,0 +1,576 @@
+import { CeloContractName } from '@celo/protocol/lib/registry-utils'
+import {
+  assertContainSubset,
+  assertEqualBN,
+  assertRevert,
+  timeTravel,
+} from '@celo/protocol/lib/test-utils'
+import BigNumber from 'bignumber.js'
+import {
+  MockElectionContract,
+  MockElectionInstance,
+  MockGoldTokenContract,
+  MockGoldTokenInstance,
+  MockSortedOraclesContract,
+  MockSortedOraclesInstance,
+  EpochRewardsTestContract,
+  EpochRewardsTestInstance,
+  RegistryContract,
+  RegistryInstance,
+} from 'types'
+import { fromFixed, toFixed } from '@celo/utils/lib/fixidity'
+
+const EpochRewards: EpochRewardsTestContract = artifacts.require('EpochRewardsTest')
+const MockElection: MockElectionContract = artifacts.require('MockElection')
+const MockGoldToken: MockGoldTokenContract = artifacts.require('MockGoldToken')
+const MockSortedOracles: MockSortedOraclesContract = artifacts.require('MockSortedOracles')
+const Registry: RegistryContract = artifacts.require('Registry')
+
+// @ts-ignore
+// TODO(mcortesi): Use BN
+EpochRewards.numberFormat = 'BigNumber'
+
+const YEAR = new BigNumber(365 * 24 * 60 * 60)
+const SUPPLY_CAP = new BigNumber(web3.utils.toWei('1000000000'))
+
+const getExpectedTargetTotalSupply = (timeDelta: BigNumber) => {
+  const genesisSupply = new BigNumber(web3.utils.toWei('600000000'))
+  const linearRewards = new BigNumber(web3.utils.toWei('200000000'))
+  return genesisSupply
+    .plus(timeDelta.times(linearRewards).div(YEAR.times(15)))
+    .integerValue(BigNumber.ROUND_FLOOR)
+}
+
+contract('EpochRewards', (accounts: string[]) => {
+  let epochRewards: EpochRewardsTestInstance
+  let mockElection: MockElectionInstance
+  let mockGoldToken: MockGoldTokenInstance
+  let mockSortedOracles: MockSortedOraclesInstance
+  let registry: RegistryInstance
+  const nonOwner = accounts[1]
+
+  const targetVotingYieldParams = {
+    initial: toFixed(new BigNumber(1 / 20)),
+    max: toFixed(new BigNumber(1 / 5)),
+    adjustmentFactor: toFixed(new BigNumber(1 / 365)),
+  }
+  const rewardsMultiplier = {
+    max: toFixed(new BigNumber(2)),
+    adjustments: {
+      underspend: toFixed(new BigNumber(1 / 2)),
```

Comment:
Will we always want `underspend`  < 1 and `overspend` > 1? If so, should these conditions be required in the contract?

Reply:
Good call

Reply:
Actually, after thinking about this more, I think this doesn't strictly need to be true, wdyt @rcroessmann?

Reply:
I think we should enforce both. Especially the condition overspend > 1 seems necessary as the protocol could end up trying to spend more than the remaining cGLD for overspend <= 1.